client/executor/src/wasm_runtime.rs

```diff
@@ -104,87 +124,127 @@ impl RuntimesCache {
-					let host_functions_changed = cached_runtime.runtime.host_functions()
-						!= host_functions;
-					if heap_pages_changed || host_functions_changed {
-						let changed = if heap_pages_changed {
-							"heap_pages"
-						} else {
-							"host functions"
-						};
+		let runtimes = self.runtimes.upgradable_read(); // this must be released prior to calling f
+		let pos = runtimes.iter().position(|r| r.as_ref().map_or(
+			false,
+			|r| r.wasm_method == wasm_method && r.code_hash == code_hash && r.heap_pages == heap_pages
+		));
 
-						trace!(
-							target: "runtimes_cache",
-							"{} were changed. Reinstantiating the instance",
-							changed,
-						);
-						*result = create_versioned_wasm_runtime(
-							ext,
-							wasm_method,
-							heap_pages,
-							host_functions.into(),
-						);
-						if let Err(ref err) = result {
-							warn!(target: "runtimes_cache", "cannot create a runtime: {:?}", err);
-						}
+		let runtime = match pos {
+			Some(n) => runtimes[n].clone().expect("`position` only returns `Some` for entries that are `Some`"),
+			None =>  {
+				let code = match code {
+					CodeSource::Externalities => {
+						Cow::Owned(ext.original_storage(well_known_keys::CODE).ok_or(WasmError::CodeNotFound)?)
 					}
-				}
-				result
-			},
-			Entry::Vacant(v) => {
-				trace!(target: "runtimes_cache", "no instance found in cache, creating now.");
+					CodeSource::Custom(code) => {
+						Cow::Borrowed(code)
+					}
+				};
+
 				let result = create_versioned_wasm_runtime(
+					&code,
+					code_hash,
 					ext,
 					wasm_method,
 					heap_pages,
 					host_functions.into(),
+					allow_missing_func_imports,
 				);
 				if let Err(ref err) = result {
-					warn!(target: "runtimes_cache", "cannot create a runtime: {:?}", err);
+					log::warn!(target: "wasm-cache", "Cannot create a runtime: {:?}", err);
 				}
-				v.insert(result)
+				Arc::new(result?)
 			}
 		};
 
-		result.as_mut()
-			.map(|entry| (entry.runtime.as_mut(), &entry.version, code_hash))
-			.map_err(|ref e| Error::InvalidCode(format!("{:?}", e)))
-	}
+		// Rearrange runtimes by last recently used.
+		match pos {
+			Some(0) => { drop(runtimes) },
+			Some(n) => {
+				let mut runtimes = RwLockUpgradableReadGuard::upgrade(runtimes);
+				for i in (1 .. n + 1).rev() {
+					runtimes.swap(i, i - 1);
+				}
+			}
+			None => {
+				let mut runtimes = RwLockUpgradableReadGuard::upgrade(runtimes);
+				runtimes[MAX_RUNTIMES-1] = Some(runtime.clone());
+				for i in (1 .. MAX_RUNTIMES).rev() {
+					runtimes.swap(i, i - 1);
+				}
+			}
+		}
+
+		let result = {
+			// Find a free instance
+			let instance_pool = { runtime.instances.read().clone() };
+			let instance = instance_pool.iter().filter_map(|i| i.as_ref().and_then(|i| i.try_lock())).next();
+			if let Some(locked) = instance {
+				f(&**locked, runtime.version.as_ref(), ext)
+			} else {
+				// Allocate a new instance
+				drop(instance);
```

Comment:
This is needed to drop `instance_pool` borrow, right?

Reply:
I believe it is no longer needed after a recent commit where I've reworked pool access.